plutus-ledger/src/Ledger/Typed/Scripts.hs

```diff
@@ -1,45 +1,45 @@
 {-# LANGUAGE DerivingStrategies   #-}
 {-# LANGUAGE GADTs                #-}
-{-# LANGUAGE TypeApplications     #-}
 {-# LANGUAGE UndecidableInstances #-}
 
 {-# OPTIONS_GHC -Wno-missing-import-lists #-}
 
 module Ledger.Typed.Scripts
   ( module Export
-  , TypedScriptTxIn (tyTxInTxIn, tyTxInOutRef)
-  , makeTypedScriptTxIn
+  , MintingPolicy
+  , Validator
+  , PV1.ConnectionError (..)
+  , mkForwardingMintingPolicy
+  , unsafeMkTypedValidator
+  -- TODO: Don't export Plutus V1 specific code from a module that doesn't mention a plutus version
+  , PV1.ValidatorType
+  , PV1.mkTypedValidator
+  , PV1.mkTypedValidatorParam
+  , PV1.mkUntypedMintingPolicy
+  , PV1.mkUntypedValidator
   ) where
 
-import Ledger.Tx.Internal (TxIn (TxIn), TxInType (ScriptAddress))
 import Ledger.Typed.Scripts.Orphans as Export ()
+import Plutus.Script.Utils.Scripts qualified as Untyped
 import Plutus.Script.Utils.Typed as Export
-import Plutus.Script.Utils.V1.Typed.Scripts as Export
-import Plutus.V1.Ledger.Api (Datum (Datum), Redeemer (Redeemer), ToData (..))
+import Plutus.Script.Utils.V1.Typed.Scripts qualified as PV1
+import Plutus.Script.Utils.V2.Typed.Scripts qualified as PV2
+import Plutus.V1.Ledger.Api (MintingPolicy, Validator)
 
--- | A 'TxIn' tagged by two phantom types: a list of the types of the data scripts in the transaction; and the connection type of the input.
-data TypedScriptTxIn a = TypedScriptTxIn
-  { tyTxInTxIn   :: TxIn,
-    tyTxInOutRef :: TypedScriptTxOutRef a
-  }
+mkForwardingMintingPolicy :: Versioned Validator -> Versioned MintingPolicy
+mkForwardingMintingPolicy vl@(Versioned _ PlutusV1) = Versioned (PV1.mkForwardingMintingPolicy (Untyped.validatorHash vl)) PlutusV1
+mkForwardingMintingPolicy vl@(Versioned _ PlutusV2) = Versioned (PV2.mkForwardingMintingPolicy (Untyped.validatorHash vl)) PlutusV2
 
-instance Eq (DatumType a) => Eq (TypedScriptTxIn a) where
-  l == r =
-    tyTxInTxIn l == tyTxInTxIn r
-      && tyTxInOutRef l == tyTxInOutRef r
+-- | Make a 'TypedValidator' (with no type constraints) from an untyped 'Validator' script.
+unsafeMkTypedValidator :: Versioned Validator -> TypedValidator Any
+unsafeMkTypedValidator vl =
```

Comment:
Not really related to the PR but still: did we officially decide of a strategy regarding the content of `Ledger.Typed.Scripts` vs `PSU.Vx.Typed.Scripts` modules? The deployment of V2 scripts may require clarification. We have an epic about cleaning the code after the Babbage era release, maybe we need to add a review/clean up/harmonisation of these modules to it? (cc @koslambrou)

Reply:
I'd say at least keep the version specific things as much as possible in `PSU.Vx.Typed.Scripts`, and let the rest work with `Versioned` scripts. It's tempting to export "the latest version" code from `Ledger.Typed.Scripts` too, but I'm not sure we should do that, and at least it would silently break a lot of our tests at the moment. (But that's a bug really, tests that depend on something being version 1 should explicitly import from `V1` modules.)

Reply:
> I'd say at least keep the version specific things as much as possible in `PSU.Vx.Typed.Scripts`, and let the rest work with `Versioned` scripts. It's tempting to export "the latest version" code from `Ledger.Typed.Scripts` too, but I'm not sure we should do that, and at least it would silently break a lot of our tests at the moment. (But that's a bug really, tests that depend on something being version 1 should explicitly import from `V1` modules.)

Yes I was thinking of something alike, it would be quite aligned with Konstantinos' ADR about cardano API as well. We should probably stick to this behaviour in `Ledger.Test` as well (and maybe in other places, but I don't see where). At least it would give us a set of clearer rules to deal with plutus versions. This would probably imply moving Versioned to plutus-ledger as well?